source/Calamari.Aws/Integration/CloudFormation/CloudFormationObjectExtensions.cs

```diff
@@ -0,0 +1,327 @@
+        }
+
+        /// <summary>
+        /// Get the potentially useful inner web exception message if there is one.
+        /// </summary>
+        /// <param name="exception"></param>
+        /// <returns></returns>
+        public static string GetWebExceptionMessage(this AmazonServiceException exception)
+        {
+            return (exception.InnerException as WebException)?
+                   .Response?
+                   .GetResponseStream()?
+                   .Map(stream => new StreamReader(stream).ReadToEnd())
+                   .Map(message => "An exception was thrown while contacting the AWS API.\n" + message)
+                   ?? "An exception was thrown while contacting the AWS API.";
+        }
+
+        /// <summary>
+        /// Wait for a given stack to complete by polling the stack events
+        /// </summary>
+        /// <param name="clientFactory">The client factory method to use</param>
+        /// <param name="waitPeriod">The period to wait between events</param>
+        /// <param name="stack">The stack name or id to query</param>
+        /// param name="action">Callback for each event while waiting
+        /// <param name="filter">The predicate for filtering the stack events</param>
+        public static void WaitForStackToComplete(this Func<IAmazonCloudFormation> clientFactory,
+            TimeSpan waitPeriod, StackArn stack, Action<Maybe<StackEvent>> action = null, Func<StackEvent, bool> filter= null)
+        {
+            Guard.NotNull(stack, "Stack should not be null");
+            Guard.NotNull(clientFactory, "Client factory should not be null");
+            
+            var status = clientFactory.StackExists(stack, StackStatus.DoesNotExist);
+            if (status == StackStatus.DoesNotExist || status == StackStatus.Completed)
+            {
+                return;
+            }
+
+            do
+            {
+                Thread.Sleep(waitPeriod);
+                var @event = clientFactory.GetLastStackEvent(stack, filter);
+                action?.Invoke(@event);
+            } while (clientFactory.StackExists(stack, StackStatus.Completed) == StackStatus.InProgress);
+        }
+                
+        /// <summary>
+        /// Check the stack event status to detemrine whether it was successful.
+        /// http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-describing-stacks.html#w2ab2c15c15c17c11
+        /// </summary>
+        /// <param name="status">The status to check</param>
+        /// <returns>true if the status indcates a failed create or update, and false otherwise</returns>
+        public static Maybe<bool> MaybeIndicatesSuccess(this StackEvent status)
+        {
+            return status.ToMaybe().Select(x => !UnsuccessfulStackEvents.Contains(x.ResourceStatus.Value));
+        }
+
+        public static bool StackIsUnrecoverable(this StackEvent status)
+        {
+            Guard.NotNull(status, "Status should not be null");
```

Comment:
I like these guards from a dev point-of-view :) Reminds me of `assert` in ObjectiveC